src/internal/connector/onedrive/collection.go

```diff
@@ -223,57 +223,62 @@ func (oc *Collection) populateItems(ctx context.Context) {
 			defer wg.Done()
 			defer func() { <-semaphoreCh }()
 
-			// Read the item
 			var (
+				itemName = *item.GetName()
+				itemSize = *item.GetSize()
 				itemInfo details.ItemInfo
-				itemData io.ReadCloser
-				err      error
-			)
-
-			for i := 1; i <= maxRetries; i++ {
-				itemInfo, itemData, err = oc.itemReader(oc.itemClient, item)
-				if err == nil || graph.IsErrTimeout(err) == nil {
-					// retry on Timeout type errors, break otherwise.
-					break
-				}
-
-				if i < maxRetries {
-					time.Sleep(1 * time.Second)
-				}
-			}
-
-			if err != nil {
-				errUpdater(*item.GetId(), err)
-				return
-			}
-
-			var (
-				itemName string
-				itemSize int64
 			)
 
 			switch oc.source {
 			case SharePointSource:
+				itemInfo.SharePoint = sharePointItemInfo(item, *item.GetSize())
 				itemInfo.SharePoint.ParentPath = parentPathString
-				itemName = itemInfo.SharePoint.ItemName
-				itemSize = itemInfo.SharePoint.Size
 			default:
+				itemInfo.OneDrive = oneDriveItemInfo(item, *item.GetSize())
 				itemInfo.OneDrive.ParentPath = parentPathString
-				itemName = itemInfo.OneDrive.ItemName
-				itemSize = itemInfo.OneDrive.Size
 			}
 
+			// Construct a new lazy readCloser to feed to the collection consumer.
+			// This ensures that downloads won't be attempted unless that consumer
+			// attempts to read bytes.  Assumption is that kopia will check things
+			// like file modtimes before attempting to read.
 			itemReader := lazy.NewLazyReadCloser(func() (io.ReadCloser, error) {
+				// Read the item
+				var (
+					itemData io.ReadCloser
+					err      error
+				)
+
+				for i := 1; i <= maxRetries; i++ {
+					_, itemData, err = oc.itemReader(oc.itemClient, item)
+					if err == nil || graph.IsErrTimeout(err) == nil {
+						// retry on Timeout type errors, break otherwise.
+						break
+					}
+
+					if i < maxRetries {
+						time.Sleep(1 * time.Second)
+					}
+				}
+
+				// check for errors following retries
+				if err != nil {
+					errUpdater(*item.GetId(), err)
+					return nil, err
+				}
+
+				// display/log the item download
 				progReader, closer := observe.ItemProgress(ctx, itemData, observe.ItemBackupMsg, itemName, itemSize)
 				go closer()
+
+				// Item read successfully, add to collection
+				atomic.AddInt64(&itemsRead, 1)
+				// byteCount iteration
+				atomic.AddInt64(&byteCount, itemSize)
```

Comment:
Currently stuck on this, and line 293 (can't include in comment span, it's the `reportAsCompleted()`) with an async control issue.

If we run the code as-is, the loop completes and `reportAsCompleted` gets called before all items count their successes and size.  However, if we move these lines back to their original spot (272), then we get a success and byte count before we successfully read the item.  IE: any kopia assisted skip or error would cause the count/size to be off.

The waitGroup and eager downloads previously protected us from this condition.  Ideal refactor would be for stats to accrue asynchronously in a bus, rather than be bound up in the collection instance.  In the interim, a smaller hack might be to have the caller block until all items are completed (like operations does with GC and collections), and calling `reportAsCompleted` in a goroutine that's blocked on the wg.

Reply:
@vkamra Tried some async control, ended up overwhelmed with it.  Going to punt to the easy route for now and allow successes and sizes to accrue outside the lazy reader.  Opens up the possibility of mismatched success/error counts.  But we're getting into some larger design rework to handle it.  So, at least as a temporary change, this state keeps things moving.

An upcoming PR will address expired URL retries.